test/integration/job/job_test.go

```diff
@@ -494,6 +494,429 @@ func TestJobPodFailurePolicy(t *testing.T) {
+					Succeeded:   podTermination.wantSucceeded,
+					Failed:      podTermination.wantFailed,
+					Ready:       ptr.To[int32](0),
+					Terminating: ptr.To[int32](0),
+				})
+				validateIndexedJobPods(ctx, t, clientSet, jobObj, podTermination.wantActiveIndexes, podTermination.wantCompletedIndexes, podTermination.wantFailedIndexes)
+			}
+			for i := range tc.wantConditionTypes {
+				validateJobCondition(ctx, t, clientSet, jobObj, tc.wantConditionTypes[i])
+			}
+			for i := range tc.wantJobFinishedNumMetric {
+				validateCounterMetric(ctx, t, metrics.JobFinishedNum, tc.wantJobFinishedNumMetric[i])
+			}
+			validateFinishedPodsNoFinalizer(ctx, t, clientSet, jobObj)
+		})
+	}
+}
+
+// TestSuccessPolicy_GetCompleteConditionAfterAllPodsAreTerminated tests handling of pod successful.
+// When Job matches rules, Job doesn't get both SuccessCriteriaMet and Complete.
+func TestSuccessPolicy_GetCompleteConditionAfterAllPodsAreTerminated(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobSuccessPolicy, true)()
+	closeFn, resetConfig, clientSet, ns := setup(t, "success-policy-get-complete")
+	defer closeFn()
+	ctx, cancel := startJobControllerAndWaitForCaches(t, resetConfig)
+	defer cancel()
+	resetMetrics()
+
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+		Spec: batchv1.JobSpec{
+			Parallelism:    ptr.To[int32](5),
+			Completions:    ptr.To[int32](5),
+			CompletionMode: completionModePtr(batchv1.IndexedCompletion),
+			SuccessPolicy: &batchv1.SuccessPolicy{
+				Rules: []batchv1.SuccessPolicyRule{{
+					SucceededCount: ptr.To[int32](1),
+				}},
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to create Job: %v", err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      5,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	validateIndexedJobPods(ctx, t, clientSet, jobObj, sets.New(0, 1, 2, 3, 4), "", nil)
+
+	// First pod from index 0 succeeded
+	if err = setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodSucceeded, 0); err != nil {
+		t.Fatalf("Failed tring to succeess pod with index 0")
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      0,
+		Succeeded:   1,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	validateIndexedJobPods(ctx, t, clientSet, jobObj, sets.New[int](), "0", nil)
+
+	// Verify the job has "SuccessCriteriaMet" and doesn't have "Complete" since all pods are still counting.
+	validateJobCondition(ctx, t, clientSet, jobObj, batchv1.JobSuccessCriteriaMet)
+	if getJobConditionStatus(ctx, jobObj, batchv1.JobComplete) != "" {
+		t.Fatalf("Should not get Complete condition, yet")
+	}
+
+	// Verify if the Job get Complete condition after the next reconciliation.
+	validateJobCondition(ctx, t, clientSet, jobObj, batchv1.JobSuccessCriteriaMet)
+	validateJobCondition(ctx, t, clientSet, jobObj, batchv1.JobComplete)
```

Comment:
How do you guarantee that there is a reconciliation in between?

What is finishing the running pods? I suspect it's the job controller itself sending a Delete?

Reply:
That makes sense. I'll try to improve this.

Reply:
@alculquicondor As I mentioned in https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1516721260, the job controller adds Complete condition even if some running pods still remain.

So, we can not verify this case 😓 Proven experiment implementation: https://github.com/kubernetes/kubernetes/pull/123412/commits/aa0ef26953535a05b762388f85da72b9049822b9

@alculquicondor How about removing this test?

Reply:
oh, I missed this comment. Yes, let's remove it